tests/microos/toolbox.pm

```diff
@@ -0,0 +1,85 @@
+use version_utils 'is_sle_micro';
+
+
+sub run {
+    select_console 'root-console';
+    my $user = 'test_user';
+
+    # Create user and don't ask password for sudo commands
+    assert_script_run "useradd -m $user ";
+    assert_script_run "echo \"$user ALL=(ALL) NOPASSWD:ALL\" >> /etc/sudoers";
+
+    # Display help
+    record_soft_failure 'https://github.com/kubic-project/microos-toolbox/issues/22' if (script_run('toolbox -h') != 0);
+
+    record_info 'Test',                    "Run toolbox without flags";
+    assert_script_run 'toolbox id',        timeout => 180;
+    validate_script_output 'podman ps -a', sub { m/toolbox-root/ };
+    assert_script_run 'podman rm toolbox-root';
+
+    record_info 'Test',                         "Run toolbox with a given tag";
+    assert_script_run 'toolbox -t test_tag id', timeout => 180;
+    validate_script_output 'podman ps -a',      sub { m/toolbox-root-test_tag/ };
+    assert_script_run 'podman rm toolbox-root-test_tag';
+
+    record_info 'Test', "Run toolbox with a given name";
+    if (script_run('toolbox -c test_name id') == 0) {
+        validate_script_output 'podman ps -a', sub { m/test_name/ };
+        assert_script_run 'podman rm test_name';
+    } else {
+        record_soft_failure 'https://github.com/kubic-project/microos-toolbox/issues/19';
+    }
+
+    record_info 'Test', "Rootless toolbox as $user";
+    my $prefix = "runuser -l $user -c";
+    my $uid    = script_output "$prefix 'id -u'";
+    validate_script_output "$prefix 'toolbox -u id'", sub { m/uid=${uid}\(${user}\)/ }, timeout => 180;
+    die "$user shouldn't have access to /etc/passwd!" if (script_run("$prefix 'toolbox -u touch /etc/passwd'") == 0);
+    assert_script_run "$prefix \"podman rm toolbox-$user-user\"";
+
+    record_info 'Test',                               "Rootfull toolbox as $user";
+    validate_script_output "$prefix 'toolbox -r id'", sub { m/uid=0\(root\)/ };
+    die "Root should have access to /etc/passwd!" if (script_run("$prefix 'toolbox -r touch /etc/passwd'") != 0);
+    assert_script_run "podman rm toolbox-$user";
+
+
+    record_info 'Test', 'Pulling toolbox image from different registry';
+    # Switch default registries for openSUSE MicroOS and SLE Micro
+    my $registry = is_sle_micro ? 'registry.opensuse.org'   : 'registry.suse.com';
+    my $image    = is_sle_micro ? 'opensuse/toolbox:latest' : 'suse/sle-micro/5.0/toolbox:latest';
+    my $expected = is_sle_micro ? 'opensuse'                : 'sles';
+    assert_script_run 'echo -e "REGISTRY=' . $registry . '\nIMAGE=' . $image . '" > ~/.toolboxrc';
+    validate_script_output 'toolbox cat /etc/os-release', sub { m/${expected}/ }, timeout => 180;
+    assert_script_run 'podman rm toolbox-root';
+    assert_script_run 'rm ~/.toolboxrc';
+
+
+    record_info 'Test', 'Zypper tests';
+    assert_script_run 'toolbox create -t devel';
+    if (is_sle_micro) {
+        my $mountpoint = script_output 'podman mount toolbox-root-user-devel';
```

Comment:
Shouldn't this just work?

Reply:
what do you mean? I need to get the path to the container to copy the files. SLE container comes empty with no repos (as oposite to openSUSE Toolbox image, which comes with defined repos).

Reply:
Shouldn't the SLE container come with SLE Micro repos?

Reply:
No. We already asked Jiri about it and since you need a reg code to register the system this is not possible. This is the workaround that was suggested to me... I can agree it's not very elegant, but it's the only choice we have if we want to test zypper in the toolbox container...

Reply:
Is that the official documented way? If it's currently just not meant to be working, I'd argue that `zypper` in there shouldn't be tested.

Reply:
> Is that the official documented way? If it's currently just not meant to be working, I'd argue that `zypper` in there shouldn't be tested.

zypper has to work in the toolbox container. But adding the SLE Micro channel doesn't make any sense, as the RPMs from it are already installed. Adding the SLES repos, which would make sense, is not possible due to fact that SLE-Micro is a different product and thus the reg.keys don't fit. The only way to fetch compatible repos is to add the Leap 15.2 repos, but if we do that by default, then customers would expect we do L3 support for them. That's why there is no repo added by default until PM comes up with a solution. But zypper in the toolbox container has to work, there is no way around this.

Reply:
> you would need some complicated mapping in the toolbox script to copy the authentication files

The issue there is more about the `/etc/products.d/baseproduct` in the toolbox container - the "authentication files" are actually just `/etc/zypper/credentials.d/SCCcredentials` AFAIK.

> The only way to fetch compatible repos is to add the Leap 15.2 repos, but if we do that by default, then customers would expect we do L3 support for them.

Rightfully so - offering a container with a package manager inside but no repos doesn't really make sense.
What about having the Leap repos in there, but disabled by default?

Reply:
> > you would need some complicated mapping in the toolbox script to copy the authentication files
> 
> The issue there is more about the `/etc/products.d/baseproduct` in the toolbox container - the "authentication files" are actually just `/etc/zypper/credentials.d/SCCcredentials` AFAIK.

But these credentials only give access to SLE Micro repos - therefore you cannot really install much of additional software.
 
> > The only way to fetch compatible repos is to add the Leap 15.2 repos, but if we do that by default, then customers would expect we do L3 support for them.
> 
> Rightfully so - offering a container with a package manager inside but no repos doesn't really make sense.
> What about having the Leap repos in there, but disabled by default?

Well, the package manager allows to create repos. In case system is running without full internet access, it needs repos on SMT/RMT proxy anyway, then the URL cannot be known in advance.

Reply:
> > > you would need some complicated mapping in the toolbox script to copy the authentication files
> > 
> > 
> > The issue there is more about the `/etc/products.d/baseproduct` in the toolbox container - the "authentication files" are actually just `/etc/zypper/credentials.d/SCCcredentials` AFAIK.
> 
> But these credentials only give access to SLE Micro repos - therefore you cannot really install much of additional software.

Yep, but that's nothing which could be addressed by "mapping" or anything like that - it's just how the product structure + SCC is designed.

> > > The only way to fetch compatible repos is to add the Leap 15.2 repos, but if we do that by default, then customers would expect we do L3 support for them.
> > 
> > 
> > Rightfully so - offering a container with a package manager inside but no repos doesn't really make sense.
> > What about having the Leap repos in there, but disabled by default?
> 
> Well, the package manager allows to create repos. In case system is running without full internet access, it needs repos on SMT/RMT proxy anyway, then the URL cannot be known in advance.

That should work fine, as the SMT/RMT URL is part of `/etc/SUSEConnect`, which is taken from the host.

Reply:
> > > > The only way to fetch compatible repos is to add the Leap 15.2 repos, but if we do that by default, then customers would expect we do L3 support for them.
> > > 
> > > 
> > > Rightfully so - offering a container with a package manager inside but no repos doesn't really make sense.
> > > What about having the Leap repos in there, but disabled by default?
> > 
> > 
> > Well, the package manager allows to create repos. In case system is running without full internet access, it needs repos on SMT/RMT proxy anyway, then the URL cannot be known in advance.
> 
> That should work fine, as the SMT/RMT URL is part of `/etc/SUSEConnect`, which is taken from the host.

But you have absolutely no idea which repos are present there, unless you want to stick with those of SLE Micro.

I do not question that for users it would be way more convenient if they could just install more software directly from the toolbox container. However, doing any defaults instead of clearly documenting "add the repos you need" would in the end lead into more support issues than giving users full control (and making it explicit - you cannot install anything until you specify where from).

Reply:
> But you have absolutely no idea which repos are present there, unless you want to stick with those of SLE Micro.

Yes, I'm not disagreeing with that.

> I do not question that for users it would be way more convenient if they could just install more software directly from the toolbox container. However, doing any defaults instead of clearly documenting "add the repos you need" would in the end lead into more support issues than giving users full control (and making it explicit - you cannot install anything until you specify where from).

Having repos disabled by default sounds like a good way.

What the openQA test currently does is nothing a user should ever have to do. Currently it just forces SLE Micro repos into the toolbox.

Reply:
What I can do is to disable zypper test for SLE Micro case until we have a decision how to handle that use case.

Reply:
> > But you have absolutely no idea which repos are present there, unless you want to stick with those of SLE Micro.
> 
> Yes, I'm not disagreeing with that.
> 
> > I do not question that for users it would be way more convenient if they could just install more software directly from the toolbox container. However, doing any defaults instead of clearly documenting "add the repos you need" would in the end lead into more support issues than giving users full control (and making it explicit - you cannot install anything until you specify where from).
> 
> Having repos disabled by default sounds like a good way.

Which repos?

Reply:
> Which repos?

The corresponding Leap repos, like in Thorsten's comment above.

Reply:
> > Which repos?
> 
> The corresponding Leap repos, like in Thorsten's comment above.

And as I wrote, if we add that repos, independent of enabled or not, customers will expect that we support what they contain as L3, which we are not able to do.

Reply:
> And as I wrote, if we add that repos, independent of enabled or not, customers will expect that we support what they contain as L3, which we are not able to do.

My argument is that users will have to opt-in into those. They could be called `unsupported-leap-oss` or similar to make it obvious.

Reply:
Yes, but we are talking about customers who pay us for supported service. Making it easier to get unsupported is not the best way to go.

Let's stick with no repos in the container for now and for the future discuss how to move with the right set of people (including PM).

Reply:
So, finally I leave it with just `zypper lr` and it will show there are not repos, that's it. If we come up with an official way to get the repos I will add automation for it.